### ISSUE TITLE
Not using a tmp variable causes PageTable add() to fail

### ISSUE THREAD

**Issue by tuomassalo** [processwire/processwire-issues#2013]
### Short description of the issue

Behaviour of `$page->pagetable_field->add()` depends on whether a temporary variable is used:

```php
if(isset($_GET['tmp'])) {
     // this succeeds
    $tmp = addPage('c');
    $p->ptfield->add($tmp);
} else {
    // this fails!
    $p->ptfield->add(addPage('c'));
}
```

### Expected behavior

See test script below: ProcessWire would behave identically with or without `?tmp` query string variable.

### Actual behavior

With `?tmp`, the script correctly returns `1`. Without it, the `add()` fails and the script returns `0`.

### Optional: Suggestion for a possible fix

I have no idea. I'm not a PHP expert but I'm surprised this is even possible, without some very specific introspection hacks. Maybe the problem is related to the lifespan of objects, but I failed to create a minimal reproducible example without using ProcessWire.

### Steps to reproduce the issue

Save this script to `/test.php` and run it with and without `?tmp`.

```php
<?php

use ProcessWire\Page;

require('index.php');
header('Content-Type: text/plain');


////// Install fixture templates, field and pages.
////// First, clean up previous run (if any).
$pages = $wire->wire('pages');
$templates = $wire->wire('templates');
$fields = $wire->wire('fields');

// Remove pages with template 'p' or 'c'.
$pagesToRemove = $pages->find("template=p|c");
foreach ($pagesToRemove as $page) $pages->delete($page, true);

// Remove templates 'p' and 'c'.
$pTemplate = $templates->get('p');
if ($pTemplate) $templates->delete($pTemplate);

$cTemplate = $templates->get('c');
if ($cTemplate) $templates->delete($cTemplate);

// Remove fields 'ptfield' if it exists.
$field = $fields->get('ptfield');
if ($field) $fields->delete($field);

// Create templates and fields.

$pTemplate = $templates->add('p');
$pTemplate->save();

$cTemplate = $templates->add('c');
$cTemplate->save();

$field = $fields->makeItem();
$field->name = 'ptfield';
$field->label = 'ptfield';
$field->type = 'PageTable';
$field->template_id = [$cTemplate->id];
$field->parent_id = 1;
$field->inputfield = 'InputfieldPageTable';
$field->save();
$pTemplate->fields->add($field);
$pTemplate->save();


function addPage(string $template)
{
    global $wire;
    $page = new Page();
    $page->template = $template;
    $page->parent = '/';

    $page->save();

    // NB: removing this line will make the test work
    $page = $wire->wire('pages')->get($page->id);

    return $page;
}

$p = addPage('p');

if(isset($_GET['tmp'])) {
     // this succeeds
    $tmp = addPage('c');
    $p->ptfield->add($tmp);
} else {
    // this fails!
    $p->ptfield->add(addPage('c'));
}
    
$p->save();

$p = $wire->wire('pages')->get($p->id);

echo $p->ptfield->count;
```

### Setup/Environment

* ProcessWire version: latest dev
* (Optional) PHP version: 8.1 (I suspect this might be relevant!)
* (Optional) MySQL version: Mariadb 10.6

### Additional observation: if I add more lines to the `else` block, like this:
```php
// ...
} else {
    $p->ptfield->add(addPage('c')); // this fails!
    $p->ptfield->add(addPage('c')); // this fails!
    $p->ptfield->add(addPage('c')); // this fails!
}
// ...
```

... the *first* one fails, but the others succeed. (I confirmed this by adding a `title` parameter to `addPage`.)

**Comment by ryancramerdesign** [comment]
@tuomassalo That's a strange behavior, but from our standpoint there's no difference between add(addPage)) and add($tmp), so we might be looking at an odd PHP behavior or bug. If the add() method was forcing a pass-by-reference then we'd want to use a variable rather than a return value from another function, but that method does not use a reference argument. So I don't think that part is the source of anything on the PW side, since the two are equivalent from our interface to PHP. But we should look at the parts we do have sway over. And I am particularly curious about the line below though and how it changes the behavior. What is the purpose of it since you already have the $page?
```php
// NB: removing this line will make the test work
$page = $wire->wire('pages')->get($page->id);
```
For another experiment, try changing your page save call to this:
```php
wire()->pages->save($page, [ 'uncacheAll' => false ]); 
```

**Comment by tuomassalo** [comment]
I agree, it looks like a very odd PHP (compiler?) bug. Or maybe some intended PHP corner case peculiarity, since I've confirmed this bug on a wide range of PHP versions: 7.4, 8.1 and 8.4.2.

**Comment by matjazpotocnik** [comment]
@tuomassalo, what's the status of this issue?

**Comment by matjazpotocnik** [comment]
@tuomassalo 

The "bug" is caused by an interaction between PHP's expression evaluation order, ProcessWire's page caching system, and the `uncacheAll` behavior during save.

### The Execution Flow
**Case that FAILS:** `$p->ptfield->add(addPage('c'));`

PHP evaluates this left-to-right:

1. **`$p->ptfield`** — triggers lazy loading of the PageTable value
- `FieldtypePageTable::wakeupValue()` loads pages with `['cache' => false]`
- Returns an empty PageArray (no items saved yet)
- **This PageArray is stored on `$p` as a cached field value**
2. **`addPage('c')`** — creates and saves a new page
- `$page->save()` → `PagesEditor::savePageFinish()` → `$this->pages->uncacheAll($page)`
- **`uncacheAll()` iterates ALL cached pages and calls `$page->uncache()` on each**
- **This includes `$p`! So `$p->uncache()` is called, which clears `$p`'s loaded field data**
- **The PageArray that was stored on `$p` in step 1 is now orphaned**
- Then `$pages->get($page->id)` loads the page fresh from DB (different object instance)
3. **`->add($newPage)`** — adds the page to the PageArray
- The PageArray object still exists in memory (returned in step 1)
- The page IS added to this PageArray
- **But this PageArray is no longer stored on `$p`** (it was cleared by `uncache()`)
4. **`$p->save()`** — saves `$p` and its fields
- For the PageTable field, it calls `$p->get('ptfield')`
- **But `$p`'s field cache was cleared in step 2!**
- So it re-loads from DB → empty PageArray
- **The added page is LOST**

**Case that WORKS:** `$tmp = addPage('c'); $p->ptfield->add($tmp);`

1. **`addPage('c')`** — saves page, `uncacheAll` clears cache
2. **`$p->ptfield`** — loads PageTable value AFTER the uncacheAll
- PageArray is loaded fresh and stored on `$p`
3. **`->add($tmp)`** — adds to the PageArray (which is still stored on `$p`)
4. **`$p->save()`** — saves the PageArray (which has the added page) ✓

In short:

- `$p->ptfield->add(addPage('c'))` → `__get(ptfield)` runs first, then `addPage('c')`
- `$tmp = addPage('c'); $p->ptfield->add($tmp)` → `addPage('c')` runs first, then `__get(ptfield)`

### Why "First Fails, Rest Succeed"

```php
$p->ptfield->add(addPage('c')); // first - fails
$p->ptfield->add(addPage('c')); // second - succeeds  
$p->ptfield->add(addPage('c')); // third - succeeds
```
After the first call, `$p`'s cache is cleared. On the second call, `$p->ptfield` re-loads from DB (empty), but this time the PageArray is stored on `$p` AFTER the `uncacheAll` from the first call has already occurred. The timing of when `uncacheAll` runs relative to when the PageArray is loaded determines whether the added page survives.

### The Critical Code Path
From `PagesEditor::savePageFinish()` (line ~450):

`if($options['uncacheAll']) $this->pages->uncacheAll($page);`

From `Pages::uncacheAll()`:

```php
public function uncacheAll(?Page $page = null, array $options = array()) {
    return $this->cacher->uncacheAll($page, $options);
}
```

The PagesLoaderCache::uncacheAll() method (not shown but documented) removes all pages from the memory cache and by default calls $page->uncache() on each, which clears the page's loaded field data.

The fix suggested by Ryan is correct:

`wire()->pages->save($page, ['uncacheAll' => false]);`

Alternatively, the workaround of not re-fetching the page also works:

```php
function addPage(string $template) {
    $page = new Page();
    $page->template = $template;
    $page->parent = '/';
    $page->save();
    // Remove this line:
    // $page = $wire->wire('pages')->get($page->id);
    return $page;
}
```
